sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp

```diff
@@ -0,0 +1,148 @@
+//==-------------------- user_defined_reductions.hpp -----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/defines.hpp>
+#include <sycl/group_algorithm.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext::oneapi::experimental {
+
+// ---- reduce_over_group
+template <typename GroupHelper, typename T, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<std::decay_t<GroupHelper>>), T>
+reduce_over_group(GroupHelper g, T x, BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Memory = reinterpret_cast<T *>(g.get_memory().data());
```

Comment:
That group helper is a generic thing, which is used not only in this extension. It is supposed to bundle a group and some memory together. The memory there is abstract, it is just a collection of bytes, because the spec allows you to do a `reduce_over_group` for one type and then `sort_over_group` for another type using the same group helper instance, for example.

If we were to attach a type to group helper, then it won't be so reusable anymore.

Reply:
Can we make `data` return `void *` so that `static_cast` would be enough? https://godbolt.org/z/zjMTaTo7P

Reply:
`.get_memory()` returns `sycl::span` here, which is in turn a copy of `std::span`. `.data()` returns `T *` for it and if we want to have `void *` there, we will have to ask user for a `span<void, Size>`, which doesn't look good or correct.